core/trino-main/src/main/java/io/trino/operator/project/PageProcessor.java

```diff
@@ -260,6 +260,9 @@ private void updateBatchSize(int positionCount, long pageSize)
 
         private void updateRetainedSize()
         {
+            // TODO: This is an estimate without knowing anything about the SourcePage implementation details. SourcePage
+            // should expose this information directly
+            retainedSizeInBytes = Page.getInstanceSizeInBytes(page.getChannelCount());
```

Comment:
Wow. Why Page retained size makes such a big difference? These pages should be in-flight and there should be limited number of them.